### PR TITLE
CFE-2278: Introduce Split systemd units

### PR DESCRIPTION
**Pull request by nickanderson**

### CODE REVIEW

**Comment by kacf** [comment]
Looks good.

**Comment by nickanderson** [comment]
@kacf @jimis does tthis look correct for migrating to the split systemd units? Note I have only worked on enteprise hub RPM as far as the migration goes.

File: packaging/common/cfengine-hub/preinstall.sh [hunk 1]
```diff
@@ -83,6 +83,16 @@ if is_upgrade; then
...
+# CFE-2278: Migrate to split units
+if is_upgrade; then
+  if [ -e /usr/lib/systemd/system/cfengine3-web.service ]; then
+    # Disable and remove the outdated unit.
+    # It's functionality will be replaced with multiple units during postinstall.
+    /bin/systemctl disable cfengine3-web
+    rm -f /usr/lib/systemd/system/cfengine3-web.service
+  fi
+fi
+
```
Comment (kacf) [inline packaging/common/cfengine-hub/preinstall.sh]: Unneeded, the package manager will take care of this for you.

Comment (nickanderson) [inline packaging/common/cfengine-hub/preinstall.sh]: @kacf OK i removed the line to delete the unit

File: packaging/cfengine-nova/cfengine-nova.spec.in [hunk 2]
```diff
@@ -106,7 +106,15 @@ exit 0
 # Systemd units
 %defattr(644,root,root,755)
 /usr/lib/systemd/system/cfengine3.service
-/usr/lib/systemd/system/cfengine3-web.service
+/usr/lib/systemd/system/cf-apache.service
```
Comment (nickanderson) [inline packaging/cfengine-nova/cfengine-nova.spec.in]: Same applies here as to the community spec. Non hub agent packages probably don't need these services units.

**Comment by nickanderson** [comment]
rebased on top of master

File: packaging/cfengine-community/cfengine-community.spec.in [hunk 3]
```diff
@@ -106,7 +106,15 @@ rm -rf $RPM_BUILD_ROOT
 # Systemd units
 %defattr(644,root,root,755)
 /usr/lib/systemd/system/cfengine3.service
-/usr/lib/systemd/system/cfengine3-web.service
+/usr/lib/systemd/system/cf-apache.service
```
Comment (jimis) [inline packaging/cfengine-community/cfengine-community.spec.in]: I think it's unnecessary bloat. But we would need to completely change the community unit files in order to get rid of them.

File: packaging/cfengine-nova/cfengine-nova.spec.in (hunk 2 above)
Comment (jimis) [inline packaging/cfengine-nova/cfengine-nova.spec.in]: Can't we have separate split systemd unit files for hub and non-hub packages?

File: packaging/common/cfengine-non-hub/postinstall.sh [hunk 4]
```diff
@@ -71,6 +71,11 @@ case `os_type` in
     # Register CFEngine initscript, if not yet.
     #
     if [ -x /bin/systemctl ]; then
+      # Reload systemd config to pick up newly installed units
+      /bin/systemctl daemon-reload > /dev/null 2>&1
+      # Enable service units
+      /bin/systemctl enable cf-execd > /dev/null 2>&1
+      /bin/systemctl enable cf-monitord > /dev/null 2>&1
       /bin/systemctl enable cfengine3 > /dev/null 2>&1
```
Comment (jimis) [inline packaging/common/cfengine-non-hub/postinstall.sh]: I'm not sure enabling every unit separately is the systemd way to do it. Just enable the super-unit that will enable the rest.

Comment (nickanderson) [inline packaging/common/cfengine-non-hub/postinstall.sh]: @jimis This all comes from my notes. I believe I *had* to enable each of the units or they wouldn't start.

Comment (nickanderson) [inline packaging/common/cfengine-non-hub/postinstall.sh]: @jimis regarding split units between hub and non-hub, cf-execd, cf-serverd, cf-monitord and the umbrella unit would be needed for core, the rest are enterprise specific.

**Pull request merged**